### PR TITLE
Funtional metric

### PR DESCRIPTION
**Pull request by Dokholyan**
## Before submitting (checklist)

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contribution guide](https://github.com/catalyst-team/catalyst/blob/master/CONTRIBUTING.md)?
- [ ] Did you check the code style? `catalyst-make-codestyle && catalyst-check-codestyle` (`pip install -U catalyst-codestyle`).
- [ ] Did you make sure to update the docs? We use Google format for all the methods and classes.
- [ ] Did you check the docs with `make check-docs`?
- [ ] Did you write any new necessary tests?
- [ ] Did you check that your code passes the unit tests `pytest .` ? 
- [ ] Did you add your new functionality to the docs?
- [ ] Did you update the [CHANGELOG](https://github.com/catalyst-team/catalyst/blob/master/CHANGELOG.md)?

## Type of Change

- [ ] Examples / docs / tutorials / contributors update
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves an existing feature)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

PS
- [x] I know, that I could [join slack](https://join.slack.com/t/catalyst-team-core/shared_invite/zt-d9miirnn-z86oKDzFMKlMG4fgFdZafw) for pull request discussion.

### CODE REVIEW

**Comment by Scitator** [comment]
could you please merge with the `dev` branch?

File: catalyst/metrics/funtional_metric.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+        if self.cumulative_metric is None:
+            self.cumulative_metric = value
+            self.total_examples = batch_length
+        else:
+            old = (
+                self.total_examples / (self.total_examples + batch_length) * self.cumulative_metric
+            )
+            new = batch_length / (self.total_examples + batch_length) * value
+            self.cumulative_metric = old + new
+            self.total_examples += batch_length
```
Comment (Scitator) [inline catalyst/metrics/funtional_metric.py]: could we use https://github.com/catalyst-team/catalyst/blob/dev/catalyst/metrics/_additive.py#L8 here?

Comment (Scitator) [reply catalyst/metrics/funtional_metric.py]: like https://github.com/catalyst-team/catalyst/blob/dev/catalyst/metrics/_accuracy.py#L61

File: catalyst/callbacks/__init__.py [hunk 2]
```diff
@@ -26,6 +26,7 @@
     ILRUpdater,
     LRFinder,
 )
+from catalyst.callbacks.aggregation import MetricAggregationCallback
```
Comment (Scitator) [inline catalyst/callbacks/__init__.py]: to remove from this PR

File: catalyst/callbacks/aggregation.py [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+
+import torch
+
+from catalyst.core.callback import Callback, CallbackNode, CallbackOrder
+
+if TYPE_CHECKING:
+    from catalyst.core.runner import IRunner
+
+
+class MetricAggregationCallback(Callback):
```
Comment (Scitator) [inline catalyst/callbacks/aggregation.py]: to remove from this PR :)

File: catalyst/callbacks/metrics/custom.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
+from typing import Callable
+
+from catalyst.callbacks.metric import BatchMetricCallback
+from catalyst.metrics._funtional_metric import BatchFuntionalMetric
+
+
+class CustomMetricCallback(BatchMetricCallback):
```
Comment (Scitator) [inline catalyst/callbacks/metrics/custom.py:7]: could you please add it to the docs?

File: catalyst/metrics/_funtional_metric.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
+from typing import Callable, Dict
+
+import torch
+
+from catalyst.metrics import ICallbackBatchMetric
+from catalyst.metrics._additive import AdditiveValueMetric
+
+
+class BatchFuntionalMetric(ICallbackBatchMetric):
```
Comment (Scitator) [inline catalyst/metrics/_funtional_metric.py]: could you please add it to the docs?

**Review by Scitator — DISMISSED** [review]

File: docs/api/metrics.rst [hunk 6]
```diff
@@ -54,6 +54,13 @@ ConfusionMatrixMetric
     :undoc-members:
     :show-inheritance:
 
+AdditiveValueMetric
```
Comment (Scitator) [inline docs/api/metrics.rst]:
```suggestion
BatchFunctionalMetric
```

File: catalyst/callbacks/__init__.py (hunk 2 above)
Comment (Scitator) [inline catalyst/callbacks/__init__.py]:
```suggestion
```

File: docs/api/metrics.rst [hunk 7]
```diff
@@ -54,6 +54,13 @@ ConfusionMatrixMetric
     :undoc-members:
     :show-inheritance:
 
+BatchFunctionalMetric
+~~~~~~~~~~~~~~~~~~~~
```
Comment (Scitator) [inline docs/api/metrics.rst]:
```suggestion
~~~~~~~~~~~~~~~~~~~~~~~~~
```
